modules/join/src/main/java/org/apache/ignite/internal/join/rest/InitCommandHandler.java

```diff
@@ -0,0 +1,110 @@
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import org.apache.ignite.internal.join.ClusterInitializer;
+import org.apache.ignite.internal.join.Leaders;
+import org.apache.ignite.internal.rest.ErrorResult;
+import org.apache.ignite.internal.rest.netty.RestApiHttpRequest;
+import org.apache.ignite.internal.rest.netty.RestApiHttpResponse;
+import org.apache.ignite.internal.rest.routes.RequestHandler;
+import org.apache.ignite.lang.IgniteLogger;
+
+/**
+ * REST handler for the {@link InitCommand}.
+ */
+public class InitCommandHandler implements RequestHandler {
+    private static final IgniteLogger log = IgniteLogger.forClass(InitCommandHandler.class);
+
+    private final ObjectMapper objectMapper = new ObjectMapper();
+
+    private final ClusterInitializer clusterInitializer;
+
+    public InitCommandHandler(ClusterInitializer clusterInitializer) {
+        this.clusterInitializer = clusterInitializer;
+    }
+
+    @Override
+    public CompletableFuture<RestApiHttpResponse> handle(RestApiHttpRequest request, RestApiHttpResponse response) {
+        try {
+            InitCommand command = readContent(request);
+
+            if (log.isInfoEnabled()) {
+                log.info(
+                        "Received init command:\n\tMeta Storage nodes: {}\n\tCMG nodes: {}",
+                        command.metaStorageNodes(),
+                        command.cmgNodes()
+                );
+            }
+
+            return clusterInitializer.initCluster(command.metaStorageNodes(), command.cmgNodes())
+                    .thenApply(leaders -> successResponse(response, leaders))
+                    .exceptionally(e -> errorResponse(response, e));
+        } catch (Exception e) {
+            return CompletableFuture.completedFuture(errorResponse(response, e));
+        }
+    }
+
+    private InitCommand readContent(RestApiHttpRequest restApiHttpRequest) throws IOException {
+        ByteBuf content = restApiHttpRequest.request().content();
+
+        try (InputStream is = new ByteBufInputStream(content)) {
+            return objectMapper.readValue(is, InitCommand.class);
+        }
+    }
+
+    private static RestApiHttpResponse successResponse(RestApiHttpResponse response, Leaders leaders) {
+        if (log.isInfoEnabled()) {
+            log.info(
+                    "Init command executed successfully.\n\tMeta Storage leader ID: {}\n\tCMG leader ID: {}",
+                    leaders.metaStorageLeader(),
+                    leaders.cmgLeader()
+            );
+        }
+
+        response.json(Map.of(
+                "metaStorageLeaderId", leaders.metaStorageLeader(),
+                "cmgLeaderId", leaders.cmgLeader()
+        ));
+
+        return response;
+    }
+
+    private static RestApiHttpResponse errorResponse(RestApiHttpResponse response, Throwable e) {
+        if (e instanceof CompletionException) {
+            e = e.getCause();
```

Comment:
> What if e.getCause() is IgniteInternalException? Is it possible?

I think it can be any kind of exception

> If true, it'll be wrapped with Ignite(Public)Exception, isn't it?

Sorry, I don't understand the question, can you elaborate?

Reply:
Well, I mean to say that it's only valid to throw either public ignite exception or common java ones. In other words, we should prevent returning internal exceptions through our public api. But, I've missed that you've actually wrap it with APPLICATION_EXCEPTION 
`response.json(Map.of("error", new ErrorResult("APPLICATION_EXCEPTION", e.getMessage())));`
My fault.